presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -641,24 +645,59 @@ private RelationPlan planCrossJoinUnnest(RelationPlan leftPlan, Join joinNode, U
+            boolean isDuplicate = usedInputVariables.contains(inputVariable);
+            usedInputVariables.add(inputVariable);
             if (type instanceof ArrayType) {
                 Type elementType = ((ArrayType) type).getElementType();
                 if (!SystemSessionProperties.isLegacyUnnest(session) && elementType instanceof RowType) {
                     ImmutableList.Builder<VariableReferenceExpression> unnestVariableBuilder = ImmutableList.builder();
                     for (int i = 0; i < ((RowType) elementType).getFields().size(); i++) {
                         unnestVariableBuilder.add(unnestedVariablesIterator.next());
                     }
-                    unnestVariables.put(inputVariable, unnestVariableBuilder.build());
+                    // Skip adding to output of unnest node if it's a duplicate, it will be output from a projection node added on top of unnest node.
+                    if (isDuplicate) {
+                        ImmutableMap<VariableReferenceExpression, List<VariableReferenceExpression>> currentUnnestVariables = unnestVariables.build();
+                        ImmutableList<VariableReferenceExpression> unnestVariableList = unnestVariableBuilder.build();
+                        checkState(unnestVariableList.size() == currentUnnestVariables.get(inputVariable).size());
+                        IntStream.range(0, unnestVariableList.size()).boxed().forEach(index -> checkState(unnestVariableList.get(index).getType().equals(currentUnnestVariables.get(inputVariable).get(index).getType())));
+                        unnestOutputMappingBuilder.putAll(IntStream.range(0, unnestVariableList.size()).boxed().collect(toImmutableMap(index -> unnestVariableList.get(index), index -> currentUnnestVariables.get(inputVariable).get(index))));
+                    }
+                    else {
+                        unnestVariables.put(inputVariable, unnestVariableBuilder.build());
+                    }
                 }
                 else {
-                    unnestVariables.put(inputVariable, ImmutableList.of(unnestedVariablesIterator.next()));
+                    if (isDuplicate) {
+                        ImmutableMap<VariableReferenceExpression, List<VariableReferenceExpression>> currentUnnestVariables = unnestVariables.build();
+                        VariableReferenceExpression unnestVariable = unnestedVariablesIterator.next();
+                        checkState(currentUnnestVariables.get(inputVariable).size() == 1);
+                        checkState(unnestVariable.getType() == currentUnnestVariables.get(inputVariable).get(0).getType());
+                        unnestOutputMappingBuilder.put(unnestVariable, currentUnnestVariables.get(inputVariable).get(0));
+                    }
+                    else {
+                        unnestVariables.put(inputVariable, ImmutableList.of(unnestedVariablesIterator.next()));
```

Comment:
See if you can refactor this part into a separate method - legacy on/off the only difference is single var vs. multiple should be easy to merge the two cases and create a method.